### PR TITLE
CA-367979: Bugfixes to support livepatches

### PR DESCRIPTION
**Pull request by minglumlu**
This series contains a few bug fixes to support livepatch function:

1. Bugfix - Remove RebootHostOnLivePatchFailure after a completion of update
2. Return changed guidance from host.apply_updates
3. Bugfix - Add new unit test for livepatch failure case
4. Bugfix - Add RebootHost guidance wrongly when a livepatch failed
5. Bugfix - Wrong format of livepatch in returned updateinfo

### CODE REVIEW

File: ocaml/xapi/repository_helpers.ml [hunk 1]
```diff
@@ -560,14 +562,32 @@ let eval_guidance_for_one_update ~updates_info ~update ~kind
...
+            | Guidance.Recommended -> (
+              match
+                ( UpdateIdSet.mem upd_id upd_ids_of_livepatches
+                , UpdateIdSet.mem upd_id upd_ids_of_failed_livepatches
+                )
+              with
+              | _, true ->
+                  (* The update contains a failed livepatch. No guidance should be picked up. *)
+                  debug
+                    "%s doesn't contribute guidance due to a livepatch failure"
```
Comment (lindig) [inline ocaml/xapi/repository_helpers.ml:574]: `… doesn't provide guidance …`

**Review by lindig — APPROVED** [review]

File: ocaml/xapi/repository.ml [hunk 2]
```diff
@@ -889,10 +896,24 @@ let apply_updates' ~__context ~host ~updates_info ~livepatches ~acc_rpm_updates
...
-  (immediate_guidances, warnings)
+  match
+    GuidanceSet.equal
+      (GuidanceSet.of_list expected_immediate_guidances)
+      (GuidanceSet.of_list immediate_guidances)
+  with
+  | true ->
+      (immediate_guidances, warnings)
+  | false ->
+      let changed =
```
[the remark this answers is not in the document]
Comment (minglumlu) [reply ocaml/xapi/repository.ml:912]: The variable contains the guidances after changing. I think `changes` stands for the delta. But indeed the `changed` may not be good. I changed it to `return_of_immediate_guidances`.

File: ocaml/tests/test_repository_helpers.ml [hunk 3]
```diff
@@ -502,7 +502,8 @@ end)
 
 module EvalGuidanceForOneUpdate = Generic.MakeStateless (struct
   module Io = struct
-    type input_t = ((string * UpdateInfo.t) list * Update.t) * string list
+    type input_t =
+      ((string * UpdateInfo.t) list * Update.t) * (string list * string list)
```
[the remark this answers is not in the document]
Comment (minglumlu) [reply ocaml/tests/test_repository_helpers.ml:506]: Yeah. Thanks. I can have an improvement on it. It need some time. Will update later.

**Review by robhoes — APPROVED** [review]
Looks fine to me.

File: ocaml/tests/test_repository_helpers.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (minglumlu) [reply ocaml/tests/test_repository_helpers.ml]: Appended a commit to refine the input type of this unit test.

**Comment by minglumlu** [comment]
Hi @lindig @robhoes 
Thanks for your approvals. I just appended 2 fixups and 1 refinement on input type of a unit test. May I please have your reviews again?

**Comment by lindig** [comment]
✅ Looks still good to me

**Pull request merged**